language/vm/src/check_bounds.rs

```diff
@@ -23,506 +23,477 @@ impl<'a> BoundsChecker<'a> {
-                function_def.code.check_bounds((self.module, sig))
-            }
-        });
-
-        errors_type_signatures.chain(errors_code_units).collect()
-    }
-
-    #[inline]
-    fn verify_pool<Context, Item: 'a>(
-        kind: IndexKind,
-        iter: impl Iterator<Item = &'a Item>,
-        context: Context,
-    ) -> Vec<VMStatus>
-    where
-        Context: Copy,
-        Item: BoundsCheck<Context>,
-    {
-        iter.enumerate()
-            .flat_map(move |(idx, elem)| {
-                elem.check_bounds(context)
-                    .into_iter()
-                    .map(move |err| append_err_info(err, kind, idx))
-            })
-            .collect()
-    }
-}
-
-pub trait BoundsCheck<Context: Copy> {
-    fn check_bounds(&self, context: Context) -> Vec<VMStatus>;
-}
-
-#[inline]
-fn check_bounds_impl<T, I>(pool: &[T], idx: I) -> Option<VMStatus>
-where
-    I: ModuleIndex,
-{
-    let idx = idx.into_index();
-    let len = pool.len();
-    if idx >= len {
-        let status = bounds_error(I::KIND, idx, len, StatusCode::INDEX_OUT_OF_BOUNDS);
-        Some(status)
-    } else {
-        None
-    }
-}
-
-#[inline]
-fn check_code_unit_bounds_impl<T, I>(pool: &[T], bytecode_offset: usize, idx: I) -> Vec<VMStatus>
-where
-    I: ModuleIndex,
-{
-    let idx = idx.into_index();
-    let len = pool.len();
-    if idx >= len {
-        let status = bytecode_offset_err(
-            I::KIND,
-            idx,
-            len,
-            bytecode_offset,
-            StatusCode::INDEX_OUT_OF_BOUNDS,
-        );
-        vec![status]
-    } else {
-        vec![]
+        for module_handle in &self.module.module_handles {
+            self.check_module_handle(module_handle, &mut errors);
+        }
+        for struct_handle in &self.module.struct_handles {
+            self.check_struct_handle(struct_handle, &mut errors);
+        }
+        for function_handle in &self.module.function_handles {
+            self.check_function_handle(function_handle, &mut errors);
+        }
+        for field_handle in &self.module.field_handles {
+            self.check_field_handle(field_handle, &mut errors);
+        }
+        for struct_instantiation in &self.module.struct_def_instantiations {
+            self.check_struct_instantiation(struct_instantiation, &mut errors);
+        }
+        for function_instantiation in &self.module.function_instantiations {
+            self.check_function_instantiation(function_instantiation, &mut errors);
+        }
+        for field_instantiation in &self.module.field_instantiations {
+            self.check_field_instantiation(field_instantiation, &mut errors);
+        }
+        for struct_def in &self.module.struct_defs {
+            self.check_struct_def(struct_def, &mut errors);
+        }
+        for function_def in &self.module.function_defs {
+            self.check_function_def(function_def, &mut errors);
+        }
+        errors
```

Comment:
I like this approach! Struct indices and type parameters are checked in separate phases, which simplifies a lot of things. 

It would be better if you could document what properties are checked at what time here. It was unclear to me whether everything had been covered initially.

Reply:
yes I will do it and we both should do it for different parts as we "finish" them.
I'm glad that you see the simplification too, I think we could do better and plan to work on this and much more as soon as this PR lands.
At that point I plan to add a bit of documentation which is critical in terms of "specification" and testing